Add documentation (#71)
jupyterlab/jupyter-chat: docs/source/developers/developing_extensions/using-chat-extensions.md

# Using a chat extension in another extension

## Collaborative chat

The collaborative chat depends on [jupyter collaboration](https://jupyterlab-realtime-collaboration.readthedocs.io/en/latest/index.html)
to exchange the messages.

As a very brief summary, jupyter collaboration allows jupyterlab users to share a
document in real time, based on
[CRDT](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type). Any change

> made to the document is propagated to all users. These change can occur from the

Reviewer: ```suggestion
made to the document is propagated to all users. These changes can occur from the
```

frontend or from the backend. The shared document has an object representation in
Typescript (for the frontend) and in Python (for the backend). These representations
can be accessed and used by external extensions.

### Exposed token

`jupyterlab-collaborative-chat` expose several tokens that allow external extensions to
interact with.

#### IChatFactory

This token is composed of:

- `widgetConfig` object, to retrieve and change the current [settings](#chat-settings)
of all the chats
- `tracker`, a widget tracker that allows to track all the collaborative chats, and to
retrieve the current one.

```{caution}

> Currently the widget tracker only track the main area widgets, not the ones opened in

Reviewer: ```suggestion
Currently the widget tracker only tracks the main area widgets, not the ones opened in
```

the side panel.
```

#### IChatPanel

This token is a pointer to the left panel containing chats.\
It can be used to programmatically open chat in the panel for example, or to list the
opened chats.

#### IAutocompletionRegistry

This is the [autocompletion registry](#autocompletion-registry) used by the chat
widgets.

> Autocompletion commands can be added to it, and than usable from the chat input.

Reviewer: ```suggestion
Autocompletion commands can be added to it, and then be used from the chat input.
```
